contracts/signature_bridge/lib.rs

```diff
@@ -0,0 +1,197 @@
+            if !self.is_signed_by_governor(&data, &resource_params.sig) {
+                return Err(Error::InvalidSignatureFromGovernor);
+            }
+
+            if resource_params.nonce <= self.proposal_nonce
+                || self.proposal_nonce + 1048 < resource_params.nonce
+            {
+                return Err(Error::InvalidNonce);
+            }
+
+            if resource_params.function_sig != [0u8; 4] {
+                return Err(Error::InvalidFunctionSig);
+            }
+
+            // Save the info of "resource_id -> handler(contract)" in this contract.
+            self.resource_id_to_handler_address.insert(
+                resource_params.new_resource_id.to_vec(),
+                &resource_params.handler_address,
+            );
+
+            self.proposal_nonce = resource_params.nonce;
+
+            // TODO: Execute call to specific handler contract such as Token Wrapper Handler, Anchor Handler, Treasury Handler e.t.c
+
+            Ok(())
+        }
+
+        pub fn execute_proposal_with_signature(
+            &mut self,
+            data: Vec<u8>,
+            sig: Vec<u8>,
+        ) -> Result<()> {
+            if !self.is_signed_by_governor(&data, &sig) {
+                return Err(Error::InvalidSignatureFromGovernor);
+            }
+
+            // Parse resourceID from the data
+            let resource_id_bytes = &data[0..32];
+            let resource_id = element_encoder(resource_id_bytes);
+
+            // Parse chain ID + chain type from the resource ID
+            let execution_chain_id_type: u64 = get_chain_id_type(&resource_id_bytes[26..32]);
+            if compute_chain_id_type(self.chain_id, &self.chain_type) != execution_chain_id_type {
+                return Err(Error::WrongChainExecution);
+            }
+
+            let handler_address = self
+                .resource_id_to_handler_address
+                .get(resource_id.to_vec());
+
+            if handler_address.is_none() {
+                return Err(Error::InvalidResourceId);
+            }
+
+            //TODO: Execute the "proposal" in "handler" contract such as Token Wrapper Handler, Anchor Handler, Treasury Handler e.t.c
+
+            Ok(())
+        }
+
+        fn is_signed_by_governor(&mut self, data: &[u8], sig: &[u8]) -> bool {
+            let result = SignatureVerifier::verify(&self.governor, data, sig)
+                .unwrap_or_else(|error| panic!("could not verify due to: {:?}", error));
+            result
+        }
+    }
+
+    pub fn truncate_and_pad(t: &[u8]) -> Vec<u8> {
+        let mut truncated_bytes = t[..20].to_vec();
+        truncated_bytes.extend_from_slice(&[0u8; 12]);
+        truncated_bytes
+    }
+
+    pub fn element_encoder(v: &[u8]) -> [u8; 32] {
+        let mut output = [0u8; 32];
+        output.iter_mut().zip(v).for_each(|(b1, b2)| *b1 = *b2);
+        output
+    }
+
+    /// Get the `chain_id_type` from bytes array.
+    pub fn get_chain_id_type(chain_id_type: &[u8]) -> u64 {
```

Comment:
Can we maybe add these functions to webb-rs and export them so we aren't re-using this so often?

Reply:
Alright